modules/ding_config_import/ding_config_import.module

```diff
@@ -0,0 +1,169 @@
+      '#title' => t('Configuration upload'),
+      '#description' => t('This module comes with an <a href="/@url">example file</a> that can be used as an template for the upload. Note that enabled/disable modules may timeout, if it takes longer than the web-servers configuration allows.', ['@url' => $url]),
+    ],
+  ];
+
+  $form['wrapper']['upload'] = [
+    '#title' => t('Upload configurations file (yml)'),
+    '#type'  => 'file',
+  ];
+
+  $form['submit_upload'] = array(
+    '#type' => 'submit',
+    '#value' => t('Upload'),
+  );
+
+  return $form;
+}
+
+/**
+ * Form submit handler.
+ */
+function ding_config_import_admin_form_submit($form, &$form_state) {
+  $error = false;
+  $msg = t('The configuration options have been saved. You may need to clear cache to rebuild things if you have enabled/disabled modules etc.');
+  $file = $_FILES['files']['tmp_name']['upload'];
+
+  if (libraries_load('yaml')) {
+    $data = file_get_contents($file);
+    $values = Yaml::parse($data);
+    if (is_array($values)) {
+      _ding_config_import_set_values($values);
+    }
+    else {
+      $error = true;
+      $msg = t('Fail to parse the YAML file.');
+    }
+  }
+  else {
+    $error = true;
+    $msg = t('Fail to load YAML parsing library. Please contact the system administrator.');
+  }
+
+  if ($error) {
+    drupal_set_message($msg, 'error');
+  }
+  else {
+    drupal_set_message($msg, 'status');
+  }
+
+  // Ensure tmp file is removed. It contains secrets.
+  @unlink($file);
+}
+
+/**
+ * Recursive walk the array using the key as prefix.
+ *
+ * If this an multilevel array the keys are joined with "_". If the flatten
+ * parameter is FALSE.
+ *
+ * @param array $values
+ *   The values.
+ * @param string $prefix
+ *   The prefix, array key form previous recursion.
+ */
+function _ding_config_import_set_values(array $values, $prefix = '') {
+  foreach ($values as $key => $value) {
+    switch ($key) {
+      case 'modules':
+        // Disable modules.
+        if (!empty($value['disable'])) {
+          module_disable($value['disable']);
+        }
+
+        // Enabled modules.
+        if (!empty($value['enable'])) {
+          if (!module_enable($value['enable'])) {
+            // Can not throw an exception here as that would stop the other
+            // changes from being applied. So use set message directly here to
+            // notify the use that module enable failed.
+            drupal_set_message(t('Not all module was enabled check the name and the drupal logs for errors.'), 'error');
+          }
+        }
+        break;
+
+      default:
+        if (is_array($value) && _ding_config_import_is_associative($value)) {
+          if (!isset($value['array'])) {
+            _ding_config_import_set_values($value, _ding_config_import_prefix($prefix, $key));
+            break;
+          }
+        }
+
+        if (isset($value['array']) && $value['array'] == 1) {
+          unset($value['array']);
+        }
```

Comment:
```suggestion
        if (is_array($value) && _ding_config_import_is_associative($value)) {
          if (!isset($value['array']) || $value['array'] != 1  ) {
            _ding_config_import_set_values($value, _ding_config_import_prefix($prefix, $key));
            break;
          } else {
            unset($value['array']);          
          }          
        }
```

I found the multiple conditionals hard to grok. I suggest changing to this as the outer if-statement tests for associative arrays. That has to be true before checking for whatever value the `array` entry might be.

Reply:
Suggestion added